File: archiso/airootfs/usr/share/churros/control-center/widgets/card.py
import os

import gi

gi.require_version("Gtk", "4.0")

from gi.repository import Gtk


ICON_DIR = os.path.abspath(
    os.path.join(
        os.path.dirname(__file__),
        "..",
        "assets",
        "icons"
    )
)


class Card(Gtk.Button):

    def __init__(
        self,
        icon,
        title,
        subtitle
    ):

        super().__init__()

        self.add_css_class("card")

        content = Gtk.Box(
            orientation=Gtk.Orientation.HORIZONTAL,
            spacing=18
        )

        content.set_margin_top(18)
        content.set_margin_bottom(18)
        content.set_margin_start(18)
        content.set_margin_end(18)

        self.image = Gtk.Image()

        self.image.set_from_file(
            os.path.join(
                ICON_DIR,
                icon
            )
        )

        self.image.set_pixel_size(34)

        labels = Gtk.Box(
            orientation=Gtk.Orientation.VERTICAL,
            spacing=4
        )

        labels.set_valign(Gtk.Align.CENTER)

        self.title = Gtk.Label(label=title)
        self.title.set_xalign(0)
        self.title.add_css_class("card-title")

        self.subtitle = Gtk.Label(label=subtitle)
        self.subtitle.set_xalign(0)
        self.subtitle.add_css_class("card-subtitle")

        labels.append(self.title)
        labels.append(self.subtitle)

        content.append(self.image)
        content.append(labels)

        self.set_child(content)

        self.set_hexpand(True)

        self.set_size_request(
            190,
            110
        )

    def set_state(
        self,
        subtitle=None,
        icon=None
    ):

        if subtitle is not None:

            self.subtitle.set_label(
                subtitle
            )

        if icon is not None:

            self.image.set_from_file(
                os.path.join(
                    ICON_DIR,
                    icon
                )
            )

File: rust/control-center/src/widgets/card.rs
// ==========================================
// card.rs — tarjeta base del control center (port de widgets/card.py)
// ==========================================

use gtk::prelude::*;

use super::super::assets;

pub struct Card {
    pub button: gtk::Button,
    image: gtk::Image,
    pub subtitle: gtk::Label,
    pub detail: gtk::Label,
}

impl Card {
    pub fn new(icon: &str, title: &str, subtitle: &str) -> Self {
        let button = gtk::Button::new();
        button.add_css_class("card");

        let content = gtk::Box::new(gtk::Orientation::Horizontal, 18);
        content.set_margin_top(18);
        content.set_margin_bottom(18);
        content.set_margin_start(18);
        content.set_margin_end(18);

        let image = gtk::Image::from_file(assets::icon_path(icon));
        image.set_pixel_size(34);

        let labels = gtk::Box::new(gtk::Orientation::Vertical, 2);
        labels.set_valign(gtk::Align::Center);

        let title_label = gtk::Label::new(Some(title));
        title_label.set_xalign(0.0);
        title_label.add_css_class("card-title");

        let subtitle_label = gtk::Label::new(Some(subtitle));
        subtitle_label.set_xalign(0.0);
        subtitle_label.add_css_class("card-subtitle");
        subtitle_label.set_ellipsize(gtk::pango::EllipsizeMode::End);

        let detail_label = gtk::Label::new(None);
        detail_label.set_xalign(0.0);
        detail_label.add_css_class("card-detail");
        detail_label.set_visible(false);
        detail_label.set_ellipsize(gtk::pango::EllipsizeMode::End);

        labels.append(&title_label);
        labels.append(&subtitle_label);
        labels.append(&detail_label);

        content.append(&image);
        content.append(&labels);

        button.set_child(Some(&content));
        button.set_hexpand(true);
        button.set_size_request(190, 110);

        Self {
            button,
            image,
            subtitle: subtitle_label,
            detail: detail_label,
        }
    }

    pub fn set_state(&self, subtitle: Option<&str>, icon: Option<&str>) {
        if let Some(subtitle) = subtitle {
            self.subtitle.set_label(subtitle);
        }
        if let Some(icon) = icon {
            self.image.set_from_file(Some(assets::icon_path(icon)));
        }
    }

    pub fn set_detail(&self, detail: Option<&str>) {
        match detail {
            Some(d) if !d.is_empty() => {
                self.detail.set_label(d);
                self.detail.set_visible(true);
            }
            _ => {
                self.detail.set_visible(false);
            }
        }
    }
}